pipelines/idr2_preprocessing/preprocess_data.py

```diff
@@ -0,0 +1,379 @@
+
+
+#-------------------------------------------------------------------------------
+# Crosstalk Subtraction
+#-------------------------------------------------------------------------------
+if xtalk_sub:
+    # get algorithm parameters
+    globals().update(cf['algorithm']['xtalk'])
+    time = datetime.utcnow()
+    hp.utils.log("\n{}\nstarting crosstalk subtraction: {}\n".format("-"*60, time), f=lf, verbose=verbose)
+
+    # get datafiles
+    datafiles, datapols = uvt.utils.search_data(input_data_template, pols, matched_pols=False, reverse_nesting=False, flatten=True)
+
+    # write xtalk sub function
+    def xtalk_sub(j):
+        try:
+            # get datafile
+            df = datafiles[j]
+            # instantiate FRF object
+            F = hc.frf.FRFilter()
+            # load data
+            F.load_data(df)
+            # perform full time-average to get xtalk spectrum
+            F.timeavg_data(1e10, rephase=False, verbose=verbose)
+            # Delay Filter if desired
+            if dly_filter:
+                # RFI Flag
+                for k in F.avg_data.keys():
+                    # RFI Flag
+                    new_f = hq.xrfi.xrfi(F.avg_data[k], f=F.avg_flags[k], **rfi_params)
+                    F.avg_flags[k] += new_f
+                # Delay Filter
+                DF = hc.delay_filter.Delay_Filter()
+                DF.load_data(F.inp_uvdata)
+                DF.data = F.avg_data
+                DF.flags = F.avg_flags
+                DF.run_filter(**dly_params)
+                # Replace xtalk spectrum with CLEAN model
+                F.avg_data = DF.CLEAN_models
+                for k in F.avg_flags.keys():
+                    F.avg_flags[k][:] = False
+            # subtract xtalk spectrum from data
+            for k in F.data.keys():
+                F.data[k] -= F.avg_data[k]
+            # write xtalk specctrum
+            xtalk_file = xtalk_ext.join(os.path.splittext(os.path.basenname(df)))
+            xtalk_file = os.path.join(out_dir, xtalk_file)
+            F.write_data(xtalk_file, write_avg=True, overwrite=overwrite)
+            # write xtalk-subtracted data
+            out_df = os.path.join(out_dir, os.path.basename(df) + file_ext)
+            F.write_data(out_df, write_avg=False, overwrite=overwrite)
+        except:
+            err, _, tb = sys.exc_info()
+            hp.utils.log("\n{} threw {} Exception with traceback:".format(df, err), f=ef, tb=tb, verbose=verbose)
+            return 1
+
+        return 0
+
+    # run xtalk sub on each datafile
+    exit_codes = M(xtalk_sub, range(len(datafiles)))
+
+    # print to log
+    hp.utils.log("\nxtalk subtraction exit codes:\n {}".format(exit_codes), f=lf, verbose=verbose)
+
+    input_data_template += file_ext
+
+    time = datetime.utcnow()
+    hp.utils.log("\nfinished crosstalk subtraction: {}\n{}".format(time, "-"*60), f=lf, verbose=verbose)
+
```

Comment:
It's not really important right now, but it might be good in future to keep tracking of simple running-time stats for each block here. Runtimes seem to have been a useful diagnostic in the Analysis pipeline.

Reply:
for each block do you mean for each call to the run function? or each analysis choice? b/c I do the latter currently..